src/io/dma.rs

```diff
@@ -0,0 +1,209 @@
+//!
+//! Each DMA is utilized via a combination four IO registers:
+//!
+//! * **Source Address:** (`*const u32`) Where to read from. DMA0 can only read
+//!   from internal memory, the other units can read from any non-SRAM memory.
+//! * **Destination Address:** (`*mut u32`) Where to write to. DMA0/1/2 can only
+//!   write to internal memory, DMA3 can write to any non-SRAM memory.
+//! * **Word Count:** (`u16`) How many units to transfer. Despite being called
+//!   "word count" you can also use DMA to transfer half-words. DMA0/1/2 are
+//!   limited to a 14-bit counter value, DMA3 allowed the full 16-bit range to
+//!   be used for the counter. Note that even when transferring half-words you
+//!   MUST have both Source and Destination be 32-bit aligned.
+//! * **Control:** (`DMAControlSetting`) This is one of those fiddly bit-flag
+//!   registers with all sorts of settings. See the type for more info.
+//!
+//! Note that Source, Destination, and Count are all read-only, while the
+//! Control is read/write. When a DMA unit is _Enabled_ it copies the relevent
+//! Source, Destination, and Count values into its own internal registers (so a
+//! second Enable will reuse the old values). If the DMA _Repeats_ it re-copies
+//! the Count, and also the Destination if
+//! `DMADestAddressControl::IncrementReload` is configured in the Control, but
+//! not the Source.
+//!
+//! When the DMA completes the Enable bit will be cleared from the Control,
+//! unless the Repeat bit is set in the Control, in which case the Enable bit is
+//! left active and the DMA will automatically activate again at the right time
+//! (depending on the Start Timing setting). You have to manually turn off the
+//! correct bit to stop the DMA unit.
+//!
+//! ## Safety
+//!
+//! As you might have noticed by now, utilizing DMA can be very fiddly. It moves
+//! around bytes with no concern for the type system, including the `Clone` and
+//! `Copy` traits that Rust relies on. Use of DMA can be made _somewhat safe_
+//! via wrapper methods (such as those we've provided), but it's fundamentally
+//! an unsafe thing to use.
+//!
+//! ## DMA Can Cause Subtle Bugs
+//!
+//! Since the CPU is halted while DMA is active you can miss out on interrupts
+//! that should have fired. This can cause any number of unintended effects. DMA
+//! is primarily intended for loading large amounts of graphical data from ROM,
+//! or loading sound data at regulated intervals to avoid pops and crackles. It
+//! _can_ be used for general purpose bulk transfers but you are advised to use
+//! restraint.
+
+use super::*;
+
+newtype! {
+  /// Allows you to configure a DMA unit.
+  #[derive(Debug, Copy, Clone, Default, PartialEq, Eq)]
+  DMAControlSetting, u16
+}
+#[allow(missing_docs)]
+impl DMAControlSetting {
+  pub const DEST_ADDR_CONTROL_MASK: u16 = 0b11 << 5;
+  pub fn dest_address_control(self) -> DMADestAddressControl {
+    // TODO: constify
+    match self.0 & Self::DEST_ADDR_CONTROL_MASK {
```

Comment:
Wouldn't the values for the match arms be `0b00 << 5`, `0b01 << 5` etc. rather than `0`, `1`, etc.?

Reply:
UGH. Very important catch. And I even made the reverse error at first too. I have to clean all this up with another macro soon, but this is only the second register to use multi-bit parts of a register.